File: LeetCode/Java/src/problems/p0069_Sqrt/P0069.java
package problems.p0069_Sqrt;

public class P0069 {
    public int mySqrt(int x) {
        /*
         * Solution v1: Brute Force
         *
         * Runtime: ms @ (beats) %
         * Memory Usage: MB @ (beats) %
         */
        if (x <= 1) {
            return x;
        } else if (x <= 4) {
            return x / 2;
        } else if (x < 9) {
            return 2;
        }

        /*
         * Solution v1: Binary Search
         *
         * Runtime: 1 ms @ (beats) 99.99%
         * Memory Usage: 36.2 MB @ (beats) 58.12%
         */
        // int left = 1, right = x;
        // while (left <= right) {
        //     int mid = left + (right - left) / 2;
        //     // pay attention that multiple will cause overflow
        //     // thus better use division
        //     if (mid <= x / mid && (mid + 1) > x / (mid + 1)) {
        //         return mid;
        //     } else if (mid > x / mid) {
        //         right = mid - 1;
        //     } else {
        //         left = mid + 1;
        //     }
        // }
        // return left;

        /*
         * Solution v2: Brute Force
         *
         * Runtime: 217 ms @ (beats) 5.05%
         * Memory Usage: 38.5 MB @ (beats) 5.27%
         */
        // for (int i = 2; i <= x / i; i++) {
        //     if (i <= x / i && (i + 1) > x / (i + 1)) {
        //         return i;
        //     }
        // }
        // return -1;

        /*
         * (learn from Internet)
         * Solution v3: Newton
         */
        long i = x;
        while (i > x / i) {
            i = (i + x / i) / 2;
        }
        return (int) i;
    }
}

File: LeetCode/Python/solutions/p0069.py
class Solution:
    def mySqrt(self, x: int) -> int:
        #
        # Solution v1: Brute Force
        #
        # Runtime: 4908 ms @ (beats) 5.01%
        # Memory Usage: 14.1 MB @ (beats) 90.00%
        #
        # if x < 2:
        #     return x
        # elif x < 4:
        #     return 1
        # elif x < 9:
        #     return 2
        #
        # for i in range(2, x // 2):
        #     if i <= x // i and (i + 1) >= x // (i + 1):
        #         return i
        #
        # return -1

        #
        # Solution v2: Binary Search
        #
        # Runtime: 28 ms @ (beats) 96.90%
        # Memory Usage: 14.2 MB @ (beats) 70.60%
        #
        if x < 2:
            return x
        elif x < 4:
            return 1
        elif x < 9:
            return 2

        left, right = 2, x // 2

        while left < right:
            mid = left + (right - left) // 2
            if mid <= x // mid and (mid + 1) > x // (mid + 1):
                return mid
            elif mid > x // mid:
                right = mid
            else:
                left = mid

        return left
